### PR TITLE
Fixed uninstall a running binary failed on Windows

### PR DESCRIPTION
**Pull request by linyihai**
### What does this PR try to resolve?
 Fixes https://github.com/rust-lang/cargo/issues/3364

The problem reproduce when you try to uninstall a running binary and it will failed on Windows,  this is because that cargo had already remove the installed binary tracker information in disk,  and next to remove the running binary but it is not allowed in Windows.  And to the worst,  you can not uninstall the binary already and only to reinstall it by the `--force` flag.

### How to solve the issue?
This PR try to make the uninstall a binary more reasonable that  only after remove the binary sucessfully then remove the tracker information on binary and make the remove binary one by one.

### How should we test and review this PR?
Add testcase https://github.com/rust-lang/cargo/pull/13053/commits/0fd4fd357b440d4155661c3b770ed8ddee43d13e
- install the `foo`
- run `foo` in another thread.
- try to uninstall running `foo` and only failed in Windows.
- wait the `foo` finish, uninstall `foo`
- install the `foo`

### CODE REVIEW

File: src/cargo/ops/cargo_uninstall.rs [hunk 1]
```diff
@@ -146,11 +146,13 @@ fn uninstall_pkgid(
...
     }
-    tracker.save()?;
+
     for bin in to_remove {
         config.shell().status("Removing", bin.display())?;
         paths::remove_file(bin)?;
     }
 
+    // Only Save the tracker when remove Bin successfully.
+    tracker.save()?;
```
Comment (epage) [inline src/cargo/ops/cargo_uninstall.rs:156]: This misses the following part of the suggested fix

> For paths::remove_file() and tracker.remove() should go hand in hand.

Or in other words
- The old behavior was to stop tracking binaries, leaking them
- The new approach is to remove binaries but we still track them

The proposed behavior in the issue is to keep the filesystem and the tracker in sync.

Comment (linyihai) [reply src/cargo/ops/cargo_uninstall.rs]: > The new approach is to remove binaries but we still track them

If I'm not mistaken: the `the approach` is means what [this commit](09be0eef54a1e66d38012f1c1e7fd3db424e9938) had done.

Comment (linyihai) [reply src/cargo/ops/cargo_uninstall.rs]: > The proposed behavior in the issue is to keep the filesystem and the tracker in sync.
I try to solving the problem in next commit

File: tests/testsuite/install.rs [hunk 2]
```diff
@@ -2507,3 +2507,81 @@ fn install_incompat_msrv() {
...
+        .run();
+    assert_has_installed_exe(cargo_home(), "foo");
+
+    use cargo_util::ProcessBuilder;
+    use std::thread;
+    let foo_bin = cargo_test_support::install::cargo_home()
+        .join("bin")
+        .join(cargo_test_support::install::exe("foo"));
+
+    let t = thread::spawn(|| ProcessBuilder::new(foo_bin).exec().unwrap());
```
[the remark this answers is not in the document]
Comment (linyihai) [reply tests/testsuite/install.rs:2562]: it seems `cargo_test_support::process` do not work in a thread,  i take your suggestion, replace the `ProcessBuilder::new` with `cargo_test_support::process`, then the complier complains: 

```
running `C:\workspace\cargo\target\debug\cargo.exe install foo`
thread '<unnamed>' panicked at crates\cargo-test-support\src\paths.rs:95:20:
Tests must use the `#[cargo_test]` attribute in order to be able to use the crate root.
running `C:\workspace\cargo\target\debug\cargo.exe uninstall foo`
```

File: tests/testsuite/install.rs [hunk 3]
```diff
@@ -2507,3 +2508,112 @@ fn install_incompat_msrv() {
...
+    cargo_home().join(".crates2.json")
+}
+
+fn load_crates1() -> toml::Value {
+    toml::from_str(&fs::read_to_string(v1_path()).unwrap()).unwrap()
+}
+
+fn load_crates2() -> serde_json::Value {
+    serde_json::from_str(&fs::read_to_string(v2_path()).unwrap()).unwrap()
+}
```
Comment (weihanglo) [inline tests/testsuite/install.rs:2526]: nit: Could we inline these functions in `assert_tracker_noexistence`? They doesn't seem to be shared elsewhere and scattering a single logic to multiple places make it a bit harder to read.

File: tests/testsuite/install.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (weihanglo) [reply tests/testsuite/install.rs]: Good catch! `global_root` is in use in `process` so you can do that.

**Comment by linyihai** [comment]
> We're about to merge this. Let me know if you want to make more change or I'll merge it as-is.

let me make some change according you suggestion.

**Review by weihanglo — APPROVED** [review]
Thank you for the contribution!

**Pull request merged**